File: 2.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head *ListNode
	var tmp *ListNode

	tmp1 := l1
	tmp2 := l2
	i := false
	j := 0

	for {
		if tmp1 == nil || tmp2 == nil {
			break
		}
		j = 0
		if i {
			j = 1
			i = false
		}
		j = tmp1.Val + tmp2.Val + j
		if j >= 10 {
			j = j - 10
			i = true
		}
		if head == nil {
			head = &ListNode{
				Val: j,
			}
			tmp = head
		} else {
			node := &ListNode{
				Val: j,
			}

			tmp.Next = node
			tmp = node
		}
		tmp1 = tmp1.Next
		tmp2 = tmp2.Next
	}

	for tmp1 != nil {
		j = 0
		if i {
			j = 1
			i = false
		}
		j = tmp1.Val + j
		if j >= 10 {
			j = j - 10
			i = true
		}
		if head == nil {
			head = &ListNode{
				Val: j,
			}
			tmp = head
		} else {
			node := &ListNode{
				Val: j,
			}

			tmp.Next = node
			tmp = node
		}
		tmp1 = tmp1.Next
	}

	for tmp2 != nil {
		j := 0
		if i {
			j = 1
			i = false
		}
		j = tmp2.Val + j
		if j >= 10 {
			j = j - 10
			i = true
		}
		if head == nil {
			head = &ListNode{
				Val: j,
			}
			tmp = head
		} else {

			node := &ListNode{
				Val: j,
			}

			tmp.Next = node
			tmp = node
		}
		tmp2 = tmp2.Next
	}

	if i {
		node := &ListNode{
			Val: 1,
		}
		tmp.Next = node
		tmp = node
	}

	return head
}

File: c/src/main/2-add-two-numbers/2.c
#include <stdio.h>
#include <stdlib.h>
#include "2.h"
#include "../../utils.h"

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
    INIT_NODE(head, 0);

    int plus = 0;
    struct ListNode *pp = head;
    int _val;
    while (l1 != NULL && l2 != NULL) {
        _val = l1->val + l2->val + plus;
        HANDLE_VAL(_val);

        INIT_NODE(_new, _val);

        pp->next = _new;
        pp = _new;

        l1 = l1->next;
        l2 = l2->next;
    }

    HANDLE_X(l1);
    HANDLE_X(l2);

    if (plus) {
        INIT_NODE(_new, 1);

        pp->next = _new;
    }

    return head->next;
}